chain/index/api.go

```diff
@@ -0,0 +1,167 @@
+	var isNullRound bool
+	err := si.stmts.hasNullRoundAtHeightStmt.QueryRowContext(ctx, epoch).Scan(&isNullRound)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to check if null round exists at height %d: %w", epoch, err)
+	}
+	if !isNullRound {
+		return nil, xerrors.Errorf("index corruption: height %d should be a null round but is not", epoch)
+	}
+
+	return nil, nil
+}
+
+func (si *SqliteIndexer) getTipsetCountsAtHeight(ctx context.Context, height abi.ChainEpoch) (revertedCount, nonRevertedCount int, err error) {
+	err = si.stmts.countTipsetsAtHeightStmt.QueryRowContext(ctx, height).Scan(&revertedCount, &nonRevertedCount)
+	if err != nil {
+		if err == sql.ErrNoRows {
+			// No tipsets found at this height
+			return 0, 0, nil
+		}
+		return 0, 0, xerrors.Errorf("failed to query tipset counts: %w", err)
+	}
+
+	return revertedCount, nonRevertedCount, nil
+}
+
+func (si *SqliteIndexer) ChainValidateIndex(ctx context.Context, epoch abi.ChainEpoch, backfill bool) (*types.IndexValidation, error) {
+	if !si.started {
+		return nil, xerrors.Errorf("ChainValidateIndex can only be called after the indexer has been started")
+	}
+	si.writerLk.Lock()
+	defer si.writerLk.Unlock()
+
+	if err := si.sanityCheckBackfillEpoch(ctx, epoch); err != nil {
+		return nil, err
+	}
+
+	var isIndexEmpty bool
+	err := si.stmts.isIndexEmptyStmt.QueryRowContext(ctx).Scan(&isIndexEmpty)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to check if index is empty: %w", err)
+	}
+
+	// fetch the tipset at the given epoch on the canonical chain
+	expectedTs, err := si.cs.GetTipsetByHeight(ctx, epoch, nil, true)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to get tipset by height: %w", err)
+	}
+	expectedTsKeyCid, err := expectedTs.Key().Cid()
+	if err != nil {
+		return nil, xerrors.Errorf("failed to get tipset key cid: %w", err)
+	}
+
+	// Canonical chain has a null round at the epoch -> return if index is empty otherwise validate
+	if expectedTs.Height() != epoch { // Canonical chain has a null round at the epoch
+		if isIndexEmpty {
+			return nil, nil
+		}
+		return si.validateNullRound(ctx, epoch)
+	}
+
+	// if the index is empty -> short-circuit and simply backfill if applicable
+	if isIndexEmpty {
+		return si.backfillMissingTipset(ctx, expectedTs, backfill)
+	}
+
+	revertedCount, nonRevertedCount, err := si.getTipsetCountsAtHeight(ctx, epoch)
+	if err != nil {
+		if err == sql.ErrNoRows {
+			return si.backfillMissingTipset(ctx, expectedTs, backfill)
+		}
+		return nil, xerrors.Errorf("failed to get tipset counts at height: %w", err)
+	}
+	if revertedCount == 0 && nonRevertedCount == 0 {
+		return si.backfillMissingTipset(ctx, expectedTs, backfill)
+	}
+
+	if revertedCount > 0 && nonRevertedCount == 0 {
+		return nil, xerrors.Errorf("index corruption: height %d only has reverted tipsets", epoch)
+	}
+
+	if nonRevertedCount > 1 {
+		return nil, xerrors.Errorf("index corruption: height %d has multiple non-reverted tipsets", epoch)
+	}
```

Comment:
this might be more expressive as a switch block, take it or leave it

```suggestion
	switch {
	case revertedCount == 0 && nonRevertedCount == 0:
		return si.backfillMissingTipset(ctx, expectedTs, backfill)

	case revertedCount > 0 && nonRevertedCount == 0:
		return nil, xerrors.Errorf("index corruption: height %d only has reverted tipsets", epoch)

	case nonRevertedCount > 1:
		return nil, xerrors.Errorf("index corruption: height %d has multiple non-reverted tipsets", epoch)
	}
```